lib/charms/operator_libs_linux/v0/grub.py

```diff
@@ -29,4 +72,300 @@
 # to 0 if you are raising the major API version
 LIBPATCH = 1
 
-# TODO: add your code here! Happy coding!
+GRUB_DIRECTORY = Path("/etc/default/grub.d/")
+CHARM_CONFIG_PREFIX = "90-juju"
+GRUB_CONFIG = GRUB_DIRECTORY / "95-juju-charm.cfg"
+CONFIG_HEADER = f"""# This config file was produced by GRUB lib v{LIBAPI}.{LIBPATCH}.
+# https://charmhub.io/operator-libs-linux/libraries/grub
+"""
+FILE_LINE_IN_DESCRIPTION = "#   {path}"
+CONFIG_DESCRIPTION = """
+# This file represents the output of the GRUB lib, which can combine multiple
+# configurations into a single file like this.
+#
+# Original files:
+{configs}
+#
+# If you change this file, run 'update-grub' afterwards to update
+# /boot/grub/grub.cfg.
+# For full documentation of the options in this file, see:
+#   info -f grub -n 'Simple configuration'
+"""
+
+
+class ValidationError(ValueError):
+    """Exception representing value validation error."""
+
+    def __init__(self, key: str, message: str) -> None:
+        super().__init__(message)
+        self.key = key
+        self.message = message
+
+    def __str__(self) -> str:
+        """Return string representation of error."""
+        return self.message
+
+
+class IsContainerError(Exception):
+    """Exception if local machine is container."""
+
+
+class ApplyError(Exception):
+    """Exception if applying new config failed."""
+
+
+def _parse_config(stream: io.TextIOWrapper) -> Dict[str, str]:
+    """Parse config file lines."""
+    config = {}
+    for line in stream:
+        line = line.strip()
+        if not line or line.startswith("#"):
+            logger.debug("skipping line `%s`", line)
+            continue
+
+        key, value = line.split("=", 1)
+        if key in config:
+            logger.warning("key %s is duplicated in config", key)
+
+        config[key], *_ = shlex.split(value)
```

Comment:
It [sounds like](https://www.gnu.org/software/grub/manual/grub/html_node/Simple-configuration.html#Simple-configuration) `key=foo bar` is not valid? It would have to be quoted in that case, right.

I think we should issue a warning if we get things like `key=foo bar` or `key="foo bar" "baz"`, that is, more than one value. Or maybe it should even be an exception?

```python
    values = shlex.split(value)
    if len(values) > 1:
        logger.warning("key %s has more than one value", key)
    config[key] = values[0]
```

Reply:
Might also be worth sending a bunch of existing GRUB config file examples through this and see that it parses them as expected.

Reply:
I'm no expert on GRUB, but are we also going to run into trouble if the GRUB config uses scripting features? https://www.gnu.org/software/grub/manual/grub/html_node/Shell_002dlike-scripting.html

Reply:
You cannot do `GRUB_TERMINAL_INPUT=console serial` or `GRUB_TERMINAL_INPUT="console" "serial"`, because `grub-mkconfig` will failed with following error:
```bash
$ sudo grub-mkconfig -o /tmp/test
Sourcing file `/etc/default/grub'
Sourcing file `/etc/default/grub.d/50-cloudimg-settings.cfg'
Sourcing file `/etc/default/grub.d/95-juju-charm.cfg'
/usr/sbin/grub-mkconfig: 14: /etc/default/grub.d/95-juju-charm.cfg: serial: not found
```
That's why I completely ignore other values, since there should never be such situation. At least the lib will never created it, only some user can do it manually. Do you think we should check such situation? If yes I think we should raise and exception.

> I'm no expert on GRUB, but are we also going to run into trouble if the GRUB config uses scripting features? 

Me neither, but I think that this lib do not need to support it. Maybe someone will come and request it as feature, but for now it's ok to not supported. What do you say?

Reply:
Yeah, I think we should raise an exception if `shlex.split()` returns more than one value then.

Re scripting -- sounds good to ignore that and not support it for now.